### PR TITLE
Azure table storage configuration provider

### PR DESCRIPTION
**Pull request by hiraldesai**
Summary of the changes
 - Adding Azure Table Storage Configuration provider
 - Adding related tests for it

Addresses #1338

### CODE REVIEW

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+{
+    /// <summary>
+    /// Configuration extensions for AzureTableStorage configuration provider
+    /// </summary>
+    public static class AzureTableStorageConfigurationExtensions
+    {
+        /// <summary>
+        /// Adds AzureTableStorage configuration to the configuration builder
+        /// </summary>
+        /// <param name="configurationBuilder"></param>
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationExtensions.cs]: Nit: missing description. See for example https://github.com/aspnet/Extensions/blob/7fd2ab9f058570ce241a19cb12bdf8d13151a861/src/Configuration/Config.CommandLine/src/CommandLineConfigurationExtensions.cs#L19

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        public static IConfigurationBuilder AddAzureTableStorage(this IConfigurationBuilder configurationBuilder, string connectionString, string tableName = null, string partitionKey = null)
+        {
+            if (string.IsNullOrWhiteSpace(connectionString))
+            {
+                throw new ArgumentException($"{nameof(connectionString)} cannot be null or white space.", nameof(connectionString));
+            }
+
+            if (string.IsNullOrWhiteSpace(tableName))
+            {
+                tableName = "AzureTableStorageConfiguration";
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationExtensions.cs]: I would move these strings as constants to top of file.
```c#
private const string DefaultTableName = "AzureTableStorageConfiguration";
private const string DefaultPartitionKey = "AzureTableStorageConfigurationPartition";
```

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureTableStorage
+{
+    internal class AzureTableStorageConfigurationProvider : ConfigurationProvider
+    {
+        private readonly ITableStore<ConfigurationEntry> configurationEntityStore;
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs]: Nit: naming -- `_configurationEntityStore` (so leading `_` for fields).
Other places as well.

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            this.configurationEntityStore = configurationEntityStore ?? throw new ArgumentNullException(nameof(configurationEntityStore));
+            this.tableName = tableName;
+            this.partitionKey = partitionKey;
+        }
+
+        public override void Load() => LoadAsync().ConfigureAwait(false).GetAwaiter().GetResult();
+
+        private async Task LoadAsync()
+        {
+            var allItems = await configurationEntityStore.GetAllByPartitionKey(tableName, partitionKey);
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs]: Do we need support for cancellation?

See for reference https://github.com/aspnet/Extensions/blob/7fd2ab9f058570ce241a19cb12bdf8d13151a861/src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs#L45

File: src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureTableStorage
+{
+    internal class ConfigurationTableStore : ITableStore<ConfigurationEntry>
+    {
+        private const int RetryBackOffSeconds = 1;
+        private const int RetryMaxAttempts = 3;
+        private readonly CloudTableClient cloudTableClient;
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs]: Nit: naming `_`

```suggestion
        private readonly CloudTableClient _cloudTableClient;
```

File: src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+
+            // Create table client
+            cloudTableClient = cloudStorageAccount.CreateCloudTableClient();
+            cloudTableClient.DefaultRequestOptions = requestOptions;
+        }
+
+        public async Task<IEnumerable<ConfigurationEntry>> GetAllByPartitionKey(string tableName, string partitionKey)
+        {
+            if (string.IsNullOrWhiteSpace(tableName))
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs]: It's an internal class and `tableName` and `partitionKey` are already verifyed in the callee of this method. So these checks could be omitted (maybe just `Debug.Assert` them).
On the other side it is more safe to validate them. I don't have any strong opinion on this.

File: src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            }
+
+            if (cloudTableClient == null)
+            {
+                throw new Exception($"{nameof(cloudTableClient)} hasn't been initialized correctly.");
+            }
+
+            var cloudTable = cloudTableClient.GetTableReference(tableName);
+
+            await cloudTable.CreateIfNotExistsAsync();
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs]: Cancellation? (see above)

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs (hunk 4 above)
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs]:
```suggestion
            var allItems = await configurationEntityStore.GetAllByPartitionKey(tableName, partitionKey).ConfigureAwait(false);
```

File: src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs (hunk 7 above)
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs]:
```suggestion
            await cloudTable.CreateIfNotExistsAsync().ConfigureAwait(false);
```

File: src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs (hunk 6 above)
Comment (hiraldesai) [reply src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs]: No strong opinion from me on it either. Keeping it as is.

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs (hunk 4 above)
Comment (hiraldesai) [reply src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs]: Are you suggesting to add another extension overload for `AddAzureTableStorage` to allow passing cancellation token and just pass it along from `AzureTableStorageConfigurationProvider` onto the `GetAllByPartitionKey` call (and all subsequent table storage calls within the implementation)?

Or you are suggesting a `Reload` support like the `AzureKeyVaultConfigurationProvider`?

Comment (hiraldesai) [reply src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs]: FYI - I have already done the cancellation support for table storage calls.

Comment (gfoidl) [reply src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs]: > Are you suggesting to add another extension overload

No, I just meant to make cancellation similar to how it is implemented in https://github.com/aspnet/Extensions/blob/7fd2ab9f058570ce241a19cb12bdf8d13151a861/src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs#L17 an pass the token to the table-queries.

Maybe someone from the team should comment about cancelaltion / reload.

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
+using Microsoft.Extensions.Configuration.AzureTableStorage;
+using System;
+using System.Threading;
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationExtensions.cs]: Nit: order namespaces

```c#
using System;
using System.Threading;
using Microsoft.Extensions.Configuration.AzureTableStorage;
```

File: src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.RetryPolicies;
+using Microsoft.WindowsAzure.Storage.Table;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/ConfigurationTableStore.cs]: Nit: order ns

```c#
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.RetryPolicies;
using Microsoft.WindowsAzure.Storage.Table;
```

File: src/Configuration/Config.AzureTableStorage/src/ITableStore.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
+using Microsoft.WindowsAzure.Storage.Table;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/src/ITableStore.cs]: Nit: order ns

```c#
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.WindowsAzure.Storage.Table;
```

File: src/Configuration/Config.AzureTableStorage/test/AzureTableStorageConfigurationTest.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
+using Moq;
+using Xunit;
+using Microsoft.Extensions.Configuration.AzureTableStorage;
+using System.Collections.Generic;
+using System.Linq;
+using System;
+using System.Threading;
```
Comment (gfoidl) [inline src/Configuration/Config.AzureTableStorage/test/AzureTableStorageConfigurationTest.cs]: Nit: order ns

```c#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Moq;
using Microsoft.Extensions.Configuration.AzureTableStorage;
using Xunit;
```

File: src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs (hunk 4 above)
Comment (hiraldesai) [reply src/Configuration/Config.AzureTableStorage/src/AzureTableStorageConfigurationProvider.cs]: I've added the ability for consumers to pass cancellation. I'll wait for someone else to comment on the reload support.

**Comment by ajcvickers** [comment]
@hiraldesai Thank you for making this PR. Unfortunately, an Azure Table Storage provider is something we had previously discussed and decided that we did not want to implement as part of ASP.NET Core. You might instead consider releasing this code as a NuGet package so that others can benefit from your work.

**Comment by hiraldesai** [comment]
Aww... 😞 I'd have really loved to have some bragging rights for having contributed to .Net Foundation. 😄 

Ah well, some other time.

**Comment by gfoidl** [comment]
@hiraldesai please do

> releasing this code as a NuGet package so that others can benefit from your work

(so you can also contribute to the .NET ecosystem :wink:)

**Comment by anurse** [comment]
> I'd have really loved to have some bragging rights for having contributed to .Net Foundation.

We have some issues marked that we've marked "up-for-grabs" that you could look at! If something is marked with the "up-for-grabs" label, it generally means:

1. We have decided we do want to take this change
1. We think that an external contributor could contribute here (it's not super deep in the bowels of the stack ;))
1. We're willing to help get you up to speed

Here are some GitHub issue searches that can find you "up-for-grabs" issues:

* AspNetCore: https://github.com/aspnet/AspNetCore/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aup-for-grabs
* Extensions: https://github.com/aspnet/Extensions/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aup-for-grabs
* EntityFrameworkCore: https://github.com/aspnet/EntityFrameworkCore/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aup-for-grabs

Even better if you can find an issue that is in a "3.0.0-[something]" milestone because it helps us get more awesomeness in to the upcoming release!

Feel free to comment on an issue that you're interested and we'll be happy to help!

**Pull request closed without merging**